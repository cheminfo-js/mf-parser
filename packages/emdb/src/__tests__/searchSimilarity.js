'use strict';

const DBManager = require('..');

describe('test searchSimilarity', () => {
    it('should find one result with bad distribution', () => {
        let dbManager = new DBManager();
        dbManager.loadTest();
        let results = dbManager.searchSimilarity(120, { x: [120], y: [1] }, {
            precision: 1000,
            widthBottom: 0.05,
            widthTop: 0.01,
            common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
        });
        expect(results.test).toMatchSnapshot();
        expect(results.test[0].ms.similarity.value).toBeCloseTo(0.898, 2);
    });

    it('should find one result with bad bad distribution, small window', () => {
        let dbManager = new DBManager();
        dbManager.loadTest();
        let results = dbManager.searchSimilarity(120, { x: [120, 121], y: [1, 1] }, {
            precision: 1000,
            from: 120 - 0.5,
            to: 120 + 0.5,
            widthBottom: 0.1,
            widthTop: 0.1,
            common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
        });
        expect(results.test[0].ms.similarity.value).toBe(1);
    });

    it('should find one result with bad bad distribution, large window', () => {
        let dbManager = new DBManager();
        dbManager.loadTest();
        let results = dbManager.searchSimilarity(120, { x: [120], y: [1] }, {
            precision: 1000,
            from: 120 - 0.5,
            to: 120 + 2.5,
            widthBottom: 0.1,
            widthTop: 0.1,
            common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
        });
        expect(results.test[0].ms.similarity.value).toBeCloseTo(0.895, 2);
    });

    it('should find one result with bad bad distribution, large window huge width ', () => {
        let dbManager = new DBManager();
        dbManager.loadTest();
        let results = dbManager.searchSimilarity(120, { x: [120], y: [1] }, {
            precision: 1000,
            from: 120 - 0.5,
            to: 120 + 2.5,
            widthBottom: 5,
            widthTop: 5,
            common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
        });
        expect(results.test[0].ms.similarity.value).toBe(1);
    });

    it('should find one result with overlap', () => {
        let dbManager = new DBManager();
        dbManager.loadTest();
        let results = dbManager.searchSimilarity(120, { x: [120], y: [1] }, {
            precision: 1000,
            widthBottom: 0.05,
            widthTop: 0.01,
            from: 120 - 0.5,
            to: 120 + 2.5,
            common: 'first', // 'first', 'second', 'both' (or true) or 'none' (or undefined)
        });
        expect(results.test[0].ms.similarity.value).toBe(1);
    });

    it('should find one result with good distribution', () => {
        let dbManager = new DBManager();
        dbManager.loadTest();
        let results = dbManager.searchSimilarity(120, { x: [120, 121], y: [1, 0.11] }, {
            precision: 1000,
            widthBottom: 0.05,
            widthTop: 0.01,
            common: undefined, // 'first', 'second', 'both' (or true) or 'none' (or undefined)
        });
        expect(results.test[0].ms.similarity.value).toBeCloseTo(0.995, 2);
    });

});
