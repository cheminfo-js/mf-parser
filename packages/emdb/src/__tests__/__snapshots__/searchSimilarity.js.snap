// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test searchSimilarity should find one result with bad distribution 1`] = `
Array [
  Object {
    "atoms": Object {
      "C": 10,
    },
    "charge": 0,
    "em": 120,
    "ionization": Object {
      "charge": 1,
      "em": 0,
      "mf": "+",
    },
    "mf": "C10",
    "ms": Object {
      "charge": 1,
      "delta": -0.0005485799090649834,
      "em": 119.99945142009094,
      "ionization": "+",
      "ppm": 4.571499242208195,
      "similarity": Object {
        "difference": Array [
          Array [
            119.99945142009094,
            121.00280625516093,
            122.00616109023093,
            123.00951592530093,
            124.01287076037093,
            125.01622559544093,
          ],
          Array [
            0,
            0.09712608111285836,
            0.004727201865545468,
            0.00013634168256703127,
            0.0000025806105388331757,
            1.8607454944920632e-8,
          ],
        ],
        "experiment": Array [
          Array [
            120,
          ],
          Array [
            1,
          ],
        ],
        "theoretical": Array [
          Array [
            119.99945142009094,
            121.00280625516093,
            122.00616109023093,
            123.00951592530093,
            124.01287076037093,
            125.01622559544093,
          ],
          Array [
            0.8980077761210353,
            0.09712608111285836,
            0.004727201865545468,
            0.00013634168256703127,
            0.0000025806105388331757,
            1.8607454944920632e-8,
          ],
        ],
        "value": 0.8980077761210353,
      },
    },
    "mw": 120.10735896735248,
    "parts": Array [
      "C10",
    ],
    "unsaturation": 11,
  },
]
`;
